## OBJETIVO: MAPEAR LOS 'beer_style' de 'beer_reviews.csv' con los 'style_name' de la web de BeerAdvocates

Algunos 'beer_style' de 'beer_reviews.csv' son de hasta el 2015 y, algunos son algo distintos o han cambiado su denominacion respecto a los mostrados en para 2019 en la web de BeerAdvocates, 'style_name' de 'beer_styles_v1.csv' procedentes de:

https://www.beeradvocate.com/beer/styles/


In [1]:
# Data analysis libraries
import pandas as pd
import numpy as np

#### Leemos el fichero CSV: BEERADVOCATE RATINGS DEVUELTO POR EL PASO PREVIO '01.01-Data_Analysis.ipynb'

In [2]:
reviews_raw = pd.read_csv('./Data/beer_reviews_v1.csv', delimiter=',',\
                          dtype={'brewery_name': str, 'review_profilename': str,\
                                 'beer_style': str, 'beer_name': str})

In [3]:
# Haremos una copia para trabajar con ella
dfreviews = reviews_raw.copy()

nRow, nCol = dfreviews.shape
print('Hay',  nRow, 'filas y', nCol, 'columnas')

# Primer vistazo
dfreviews.head()

Hay 1489510 filas y 15 columnas


brewery_id                        brewery_name          review_time  \
0         147                   Stone Brewing Co.  2006-05-18 09:50:05   
1          35  Boston Beer Company (Samuel Adams)  2006-03-06 20:26:50   
2          42                    Brewery Ommegang  2006-05-18 07:05:12   
3          94              Long Trail Brewing Co.  2006-02-03 23:13:01   
4          33      Berkshire Brewing Company Inc.  2006-03-02 22:44:53   

   review_overall  review_aroma  review_appearance review_profilename  \
0             4.0           4.5                4.0             zyzygy   
1             4.0           3.5                3.5             zyzygy   
2             4.5           4.5                5.0             zyzygy   
3             4.5           4.0                4.0             zyzygy   
4             4.0           4.0                3.5             zyzygy   

               beer_style  review_palate  review_taste  \
0     American Barleywine            4.5           4.5   
1       English Brown Ale            3.0           3.5   
2    Saison Farmhouse Ale            4.5           5.0   
3                 Altbier            4.0           4.5   
4  Russian Imperial Stout            3.5           4.5   

                                       beer_name  beer_abv  beer_beerid  \
0  Stone Old Guardian Barley Wine Style Ale 2006     11.20        28687   
1                         Samuel Adams Brown Ale      5.35        27514   
2                    Hennepin (Farmhouse Saison)      7.70          141   
3                                     Double Bag      7.20          273   
4               Berkshire Russian Imperial Stout      8.50         1112   

   review_average  abv_strength  
0             4.3             4  
1             3.5             2  
2             4.7             3  
3             4.2             3  
4             3.9             3

#### Leemos el fichero CSV: BEERADVOCATE DE BEER STYLES DEVUELTO POR EL PASO PREVIO '01.02-Scraping-Additional-Beers-Data.ipynb'

In [4]:
styles_raw = pd.read_csv('./Data/beer_styles_v1.csv', delimiter=',',\
                          dtype={'family_name': str, 'style_name': str, 'style_URL': str})

In [5]:
nRowS, nColS = styles_raw.shape
print('Hay',  nRowS, 'filas y', nColS, 'columnas')

# Primer vistazo
styles_raw.head()

Hay 111 filas y 7 columnas


abv_range  abv_strength family_name ibu_range  ibu_strength  \
0   6.3-7.6%             3       Bocks     20-30             3   
1   6.6-7.9%             3       Bocks     17-27             3   
2  7.0-14.0%             4       Bocks     25-35             3   
3   6.3-8.1%             3       Bocks     20-38             3   
4   7.0-9.5%             3       Bocks     15-35             3   

                                      style_URL         style_name  
0  https://www.beeradvocate.com/beer/styles/32/        German Bock  
1  https://www.beeradvocate.com/beer/styles/35/  German Doppelbock  
2  https://www.beeradvocate.com/beer/styles/36/     German Eisbock  
3  https://www.beeradvocate.com/beer/styles/33/     German Maibock  
4  https://www.beeradvocate.com/beer/styles/92/  German Weizenbock

In [6]:
dfreviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1489510 entries, 0 to 1489509
Data columns (total 15 columns):
brewery_id            1489510 non-null int64
brewery_name          1489510 non-null object
review_time           1489510 non-null object
review_overall        1489510 non-null float64
review_aroma          1489510 non-null float64
review_appearance     1489510 non-null float64
review_profilename    1489510 non-null object
beer_style            1489510 non-null object
review_palate         1489510 non-null float64
review_taste          1489510 non-null float64
beer_name             1489510 non-null object
beer_abv              1489510 non-null float64
beer_beerid           1489510 non-null int64
review_average        1489510 non-null float64
abv_strength          1489510 non-null int64
dtypes: float64(7), int64(3), object(5)
memory usage: 170.5+ MB


In [7]:
styles_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 111 entries, 0 to 110
Data columns (total 7 columns):
abv_range       111 non-null object
abv_strength    111 non-null int64
family_name     111 non-null object
ibu_range       111 non-null object
ibu_strength    111 non-null int64
style_URL       111 non-null object
style_name      111 non-null object
dtypes: int64(2), object(5)
memory usage: 6.1+ KB


Crearemos la lista de los 111 beer styles de 2019 obtenidos de la web de BeerAdvocates, columna 'style_name'

In [9]:
bslist = list(styles_raw['style_name'])
print(len(bslist))
bslist

111


['German Bock',
 'German Doppelbock',
 'German Eisbock',
 'German Maibock',
 'German Weizenbock',
 'American Brown Ale',
 'English Brown Ale',
 'English Dark Mild Ale',
 'German Altbier',
 'American Black Ale',
 'Belgian Dark Ale',
 'Belgian Dubbel',
 'German Roggenbier',
 'Scottish Ale',
 'Winter Warmer',
 'American Amber Red Lager',
 'European Dark Lager',
 'German Märzen Oktoberfest',
 'German Rauchbier',
 'German Schwarzbier',
 'Munich Dunkel Lager',
 'Vienna Lager',
 'American Cream Ale',
 'Bière de Champagne Bière Brut',
 'Braggot',
 'California Common Steam Beer',
 'American Brut IPA',
 'American Imperial IPA',
 'American IPA',
 'Belgian IPA',
 'English India Pale Ale (IPA)',
 'New England IPA',
 'American Amber Red Ale',
 'American Blonde Ale',
 'American Pale Ale (APA)',
 'Belgian Blonde Ale',
 'Belgian Pale Ale',
 'Belgian Saison',
 'English Bitter',
 'English Extra Special Strong Bitter (ESB)',
 'English Pale Ale',
 'English Pale Mild Ale',
 'French Bière de Garde',
 'German

Existen 50 'beer_style' de 'beer_reviews_v1.csv' que SI coinciden con los 'style_name' de la web de BeerAdvocates

In [10]:
dfreviews[dfreviews['beer_style'].apply(lambda beer_style: beer_style in bslist)].head()

brewery_id                          brewery_name          review_time  \
0         147                     Stone Brewing Co.  2006-05-18 09:50:05   
1          35    Boston Beer Company (Samuel Adams)  2006-03-06 20:26:50   
4          33        Berkshire Brewing Company Inc.  2006-03-02 22:44:53   
8         113  Samuel Smith Old Brewery (Tadcaster)  2009-03-22 00:26:46   
9         113  Samuel Smith Old Brewery (Tadcaster)  2009-05-31 23:31:08   

   review_overall  review_aroma  review_appearance review_profilename  \
0             4.0           4.5                4.0             zyzygy   
1             4.0           3.5                3.5             zyzygy   
4             4.0           4.0                3.5             zyzygy   
8             5.0           4.5                4.5             zythus   
9             5.0           4.0                4.5             zythus   

               beer_style  review_palate  review_taste  \
0     American Barleywine            4.5           4.5   
1       English Brown Ale            3.0           3.5   
4  Russian Imperial Stout            3.5           4.5   
8          English Porter            5.0           5.0   
9       English Brown Ale            5.0           4.5   

                                       beer_name  beer_abv  beer_beerid  \
0  Stone Old Guardian Barley Wine Style Ale 2006     11.20        28687   
1                         Samuel Adams Brown Ale      5.35        27514   
4               Berkshire Russian Imperial Stout      8.50         1112   
8        Samuel Smith's, The Famous Taddy Porter      5.00          572   
9                   Samuel Smith's Nut Brown Ale      5.00          576   

   review_average  abv_strength  
0             4.3             4  
1             3.5             2  
4             3.9             3  
8             4.8             2  
9             4.6             2

In [11]:
dfbs_found = dfreviews[dfreviews['beer_style'] \
                           .apply(lambda beer_style: beer_style in bslist)]['beer_style'] \
                           .to_frame() \
                           .sort_values(['beer_style'], ascending=False) \
                           .drop_duplicates(subset= ['beer_style'], keep='first')

In [12]:
print(dfbs_found.shape)
dfbs_found

(50, 1)


beer_style
1382866                      Winter Warmer
553090                        Vienna Lager
57579    Scottish Gruit Ancient Herbed Ale
132165                        Scottish Ale
515534                Scotch Ale Wee Heavy
230807                            Rye Beer
870249              Russian Imperial Stout
486070                 Munich Dunkel Lager
606936                    Low Alcohol Beer
1222682                Japanese Rice Lager
543148                       Irish Red Ale
259120                     Irish Dry Stout
679451                Foreign Export Stout
1323514                   Flanders Red Ale
940402                  Flanders Oud Bruin
492575                  English Strong Ale
661818                       English Stout
928523                      English Porter
726402               English Pale Mild Ale
682869                    English Pale Ale
841768        English India Pale Ale (IPA)
1106058              English Dark Mild Ale
730145                   English Brown Ale
153958                      English Bitter
1397360                         Chile Beer
1274629       California Common Steam Beer
122303                             Braggot
1361101      Bière de Champagne Bière Brut
686616             Belgian Strong Pale Ale
1060716            Belgian Strong Dark Ale
460832                    Belgian Pale Ale
40984                          Belgian IPA
1415429                   Belgian Dark Ale
1381345                      Baltic Porter
473037                   American Wild Ale
130777                 American Strong Ale
814675                      American Stout
968945                     American Porter
63068              American Pale Wheat Ale
999150             American Pale Ale (APA)
714319                American Malt Liquor
1137587                       American IPA
681587             American Dark Wheat Ale
1374848                 American Brown Ale
857108                 American Blonde Ale
1195481                 American Black Ale
781012                 American Barleywine
10549             American Amber Red Lager
639078              American Amber Red Ale
757394              American Adjunct Lager

Existen 54 'beer_style' de 'beer_reviews.csv' que NO coinciden con los de 2019 de la web de BeerAdvocates y hay que MAPEARLOS A UNO DE ELLOS, 'style_name' de 'beer_styles_v1.csv'

In [14]:
dfreviews[dfreviews['beer_style'].apply(lambda beer_style: beer_style not in bslist)].head()

brewery_id               brewery_name          review_time  review_overall  \
2          42           Brewery Ommegang  2006-05-18 07:05:12             4.5   
3          94     Long Trail Brewing Co.  2006-02-03 23:13:01             4.5   
5          22                   Unibroue  2009-05-04 02:01:59             3.0   
6       23980         Yuengling Brewery   2010-02-25 18:01:49             4.5   
7         140  Sierra Nevada Brewing Co.  2009-03-21 18:02:53             4.0   

   review_aroma  review_appearance review_profilename  \
2           4.5                5.0             zyzygy   
3           4.0                4.0             zyzygy   
5           2.0                3.0             zythus   
6           4.0                4.0             zythus   
7           3.0                3.5             zythus   

                          beer_style  review_palate  review_taste  \
2               Saison Farmhouse Ale            4.5           5.0   
3                            Altbier            4.0           4.5   
5               Fruit Vegetable Beer            4.0           3.0   
6                               Bock            4.5           4.0   
7  Extra Special Strong Bitter (ESB)            5.0           4.5   

                               beer_name  beer_abv  beer_beerid  \
2            Hennepin (Farmhouse Saison)       7.7          141   
3                             Double Bag       7.2          273   
5                       Éphémère (Apple)       5.5         3523   
6                         Yuengling Bock       5.1        47812   
7  Sierra Nevada ESB (Early Spring Beer)       5.9        40492   

   review_average  abv_strength  
2             4.7             3  
3             4.2             3  
5             3.0             2  
6             4.2             2  
7             4.0             2

In [15]:
dfbs_not_found = dfreviews[dfreviews['beer_style'] \
                           .apply(lambda beer_style: beer_style not in bslist)]['beer_style'] \
                           .to_frame() \
                           .sort_values(['beer_style'], ascending=False) \
                           .drop_duplicates(subset= ['beer_style'], keep='first')

In [16]:
print(dfbs_not_found.shape)
dfbs_not_found

(54, 1)


beer_style
43033                              Witbier
1292135                          Wheatwine
1331595                         Weizenbock
535936                              Tripel
397193                         Smoked Beer
1374816                        Schwarzbier
231449                Saison Farmhouse Ale
794874                               Sahti
1360990                         Roggenbier
891477                           Rauchbier
61651                     Quadrupel (Quad)
563994                         Pumpkin Ale
776707                             Old Ale
601495                       Oatmeal Stout
1252299                 Märzen Oktoberfest
369086                 Munich Helles Lager
577277                    Milk Sweet Stout
1325694                Maibock Helles Bock
401029                         Light Lager
909339                    Lambic Unblended
1360358                       Lambic Fruit
357127                              Kölsch
1224710                              Kvass
534586                       Kristalweizen
1389945           Keller Bier Zwickel Bier
866442                  Herbed Spiced Beer
478339                          Hefeweizen
629730                            Happoshu
1000051                             Gueuze
1293236                               Gose
801847                     German Pilsener
1035303               Fruit Vegetable Beer
146455                                Faro
83866    Extra Special Strong Bitter (ESB)
812665                   Euro Strong Lager
529743                     Euro Pale Lager
536661                     Euro Dark Lager
690827                  English Barleywine
727519                             Eisbock
1150373                       Dunkelweizen
504062                              Dubbel
1126071            Dortmunder Export Lager
1063730                         Doppelbock
342563                      Czech Pilsener
33489                            Cream Ale
455322                                Bock
1179746                        Black & Tan
959565                      Bière de Garde
796729                  Berliner Weissbier
866527                 American Pale Lager
649440      American Double Imperial Stout
402960    American Double Imperial Pilsner
84255         American Double Imperial IPA
800205                             Altbier

#### ESTUDIO DEL MAPEO DE estilos de cervezas no encontrados

Hay 50 ENCONTRADOS + 51 NO ENCONTRADOS = 101 ESTILOS DE CERVEZA DE BeerAdvocates son los registrados en nuestro dataset de valoraciones HASTA 2015, en columna 'beer-style' de:

'beer_reviews.csv'.

Hay 111 ESTILOS DE CERVEZA INCLUIDOS EN BeerAdvocates PARA 2019 que son los mostrados en columna 'style-name' de 'beer_styles_v1.csv' obtenido haciendo web scrapping en:

https://www.beeradvocate.com/beer/styles/

EN 2019 EXISTEN 10 ESTILOS DE CERVEZA NUEVOS.

A continuación, añadiremos una Columna Nueva 'beer_advocates_style' para  registrar el mapeo con los nuevos ESTILOS DE CERVEZA DE 2019, o sea, valor correspondiente de la columna 'style-name' de 'beer_styles_v1.csv'.

#### 1.- Añadimos una Columna Nueva 'beer_advocates_style' para  registrar el mapeo de los 'beer_style' que SI COINCIDEN CON LOS 'style_name' y registramos ese mismo valor en la columna 'beer_advocates_style'

In [17]:
def mapping_beer_style_found(bs, bslist):
    if bs in bslist:
        return bs
    else:
        return ''

In [18]:
dfreviews['beer_advocates_style'] = dfreviews['beer_style']\
                                    .apply(lambda beer_style: mapping_beer_style_found(beer_style, bslist))

In [19]:
dfreviews.head(10)

brewery_id                          brewery_name          review_time  \
0         147                     Stone Brewing Co.  2006-05-18 09:50:05   
1          35    Boston Beer Company (Samuel Adams)  2006-03-06 20:26:50   
2          42                      Brewery Ommegang  2006-05-18 07:05:12   
3          94                Long Trail Brewing Co.  2006-02-03 23:13:01   
4          33        Berkshire Brewing Company Inc.  2006-03-02 22:44:53   
5          22                              Unibroue  2009-05-04 02:01:59   
6       23980                    Yuengling Brewery   2010-02-25 18:01:49   
7         140             Sierra Nevada Brewing Co.  2009-03-21 18:02:53   
8         113  Samuel Smith Old Brewery (Tadcaster)  2009-03-22 00:26:46   
9         113  Samuel Smith Old Brewery (Tadcaster)  2009-05-31 23:31:08   

   review_overall  review_aroma  review_appearance review_profilename  \
0             4.0           4.5                4.0             zyzygy   
1             4.0           3.5                3.5             zyzygy   
2             4.5           4.5                5.0             zyzygy   
3             4.5           4.0                4.0             zyzygy   
4             4.0           4.0                3.5             zyzygy   
5             3.0           2.0                3.0             zythus   
6             4.5           4.0                4.0             zythus   
7             4.0           3.0                3.5             zythus   
8             5.0           4.5                4.5             zythus   
9             5.0           4.0                4.5             zythus   

                          beer_style  review_palate  review_taste  \
0                American Barleywine            4.5           4.5   
1                  English Brown Ale            3.0           3.5   
2               Saison Farmhouse Ale            4.5           5.0   
3                            Altbier            4.0           4.5   
4             Russian Imperial Stout            3.5           4.5   
5               Fruit Vegetable Beer            4.0           3.0   
6                               Bock            4.5           4.0   
7  Extra Special Strong Bitter (ESB)            5.0           4.5   
8                     English Porter            5.0           5.0   
9                  English Brown Ale            5.0           4.5   

                                       beer_name  beer_abv  beer_beerid  \
0  Stone Old Guardian Barley Wine Style Ale 2006     11.20        28687   
1                         Samuel Adams Brown Ale      5.35        27514   
2                    Hennepin (Farmhouse Saison)      7.70          141   
3                                     Double Bag      7.20          273   
4               Berkshire Russian Imperial Stout      8.50         1112   
5                               Éphémère (Apple)      5.50         3523   
6                                 Yuengling Bock      5.10        47812   
7          Sierra Nevada ESB (Early Spring Beer)      5.90        40492   
8        Samuel Smith's, The Famous Taddy Porter      5.00          572   
9                   Samuel Smith's Nut Brown Ale      5.00          576   

   review_average  abv_strength    beer_advocates_style  
0             4.3             4     American Barleywine  
1             3.5             2       English Brown Ale  
2             4.7             3                          
3             4.2             3                          
4             3.9             3  Russian Imperial Stout  
5             3.0             2                          
6             4.2             2                          
7             4.0             2                          
8             4.8             2          English Porter  
9             4.6             2       English Brown Ale

#### 2.- MAPEOS DIRECTOS MANUALES.

Despues del estudio de las diferencias entre 'beer_style' y 'style_name' se identifica la necesidad de registrar a mano los valores de 'style_name' para un subconjunto de 11 ESTILOS DE CERVEZA 'beer_style' DE HASTA 2015 YA QUE NO SE PUEDE CREAR UN ALGORITMO PARA AUTOMATIZAR SU MAPEO por distintas causas:

2.a.- 'style_name' DE 2015 QUE HAN CAMBIADO A OTRO NOMBRE NUEVO DE 'style_name'.

2.a.1.- 'Herbed Spiced Beer' EN 2019 HA CAMBIADO A 'Herb and Spice Beer'

2.a.2.- 'Czech Pilsener' EN 2019 HA CAMBIADO A 'Bohemian Pilsener'

2.a.3.- 'German Pilsener' EN 2019 HA CAMBIADO A 'German Pilsner'

2.a.4.- 'Keller Bier Zwickel Bier' EN 2019 HA CAMBIADO A 'German Kellerbier Zwickelbier'

2.b.- EL ALGORITMO DE MAPEO COMUN posterior, mapping_beer_style_not_found('beer_style', lista_de_'style-name'), NO PUEDE MAPEARLOS BIEN PORQUE existen casos en que existe EMPATE, o sea, el mismo numero de palabras del 'beer_style' de HASTA 2015 que han sido encontradas en mas de un estilo actual 'sytle_name' de BeerAdvocates de 2019.

2.b.1.- 'Pumpkin Ale' DEBERIA SER 'Pumpkin Beer'

2.b.2.- 'Munich Helles' Lager DEBERIA SER 'German Helles'

2.b.3.- 'Maibock Helles Bock' DEBERIA SER 'German Maibock'MANO

2.b.4.- 'American Pale Lager' DEBERIA SER 'American Lager'

2.b.5.- 'Lambic Unblended' DEBERIA SER 'Belgian Lambic'(ver: https://birrapedia.com/enciclopedia-de-la-cerveza/straight--unblended--lambic/s)

2.b.6.- 'Saison Farmhouse Ale' DEBERIA SER 'Belgian Saison'

2.b.7.- 'English Barleywine' DEBERIA SER 'British Barleywine'


In [20]:
bs_to_manually_map_list = ['Herbed Spiced Beer', 'Czech Pilsener', \
                           'German Pilsener', 'Keller Bier Zwickel Bier', \
                           'Pumpkin Ale', 'Munich Helles Lager', 'Maibock Helles Bock', 'American Pale Lager', \
                           'Lambic Unblended', 'Saison Farmhouse Ale', 'English Barleywine']
bs_manually_mapped_list = ['Herb and Spice Beer', 'Bohemian Pilsener', \
                           'German Pilsner', 'German Kellerbier Zwickelbier', \
                           'Pumpkin Beer', 'German Helles', 'German Maibock', 'American Lager', \
                           'Belgian Lambic', 'Belgian Saison', 'British Barleywine']

In [22]:
def get_manually_mapped_bs(bs_to_map):
    mapped_bs = ''
    current_bs_index = 0
    for current_bs in bs_to_manually_map_list:
        if current_bs == bs_to_map:
            mapped_bs = bs_manually_mapped_list[current_bs_index]
            break
        else:
            current_bs_index += 1
    return mapped_bs

In [21]:
dfreviews[dfreviews['beer_style']\
          .apply(lambda beer_style: beer_style in bs_to_manually_map_list)]['beer_style'].count()

129262

COMPROBACION PREVIA DE LOS MAPEOS DIRECTOS MANUALES

In [23]:
for bs in bs_to_manually_map_list:
    print("'{0}' mapped to '{1}'".format(bs, get_manually_mapped_bs(bs)))

'Herbed Spiced Beer' mapped to 'Herb and Spice Beer'
'Czech Pilsener' mapped to 'Bohemian Pilsener'
'German Pilsener' mapped to 'German Pilsner'
'Keller Bier Zwickel Bier' mapped to 'German Kellerbier Zwickelbier'
'Pumpkin Ale' mapped to 'Pumpkin Beer'
'Munich Helles Lager' mapped to 'German Helles'
'Maibock Helles Bock' mapped to 'German Maibock'
'American Pale Lager' mapped to 'American Lager'
'Lambic Unblended' mapped to 'Belgian Lambic'
'Saison Farmhouse Ale' mapped to 'Belgian Saison'
'English Barleywine' mapped to 'British Barleywine'


In [24]:
for bs in bs_to_manually_map_list:
    dfreviews.loc[dfreviews['beer_style'] == bs, 'beer_advocates_style'] = \
        get_manually_mapped_bs(bs)

In [25]:
dfreviews[dfreviews['beer_advocates_style'] \
          .apply(lambda beer_advocates_style: beer_advocates_style in bs_manually_mapped_list)] \
         ['beer_advocates_style'].count()

129262

In [26]:
# SE HAN MAPEADO 129262 registros CORRECTAMENTE
dfreviews[dfreviews['beer_advocates_style'] \
          .apply(lambda beer_advocates_style: beer_advocates_style in bs_manually_mapped_list)] \
          ['beer_style'] \
          .apply(lambda beer_style: beer_style in bs_to_manually_map_list).count()

129262

#### 3.- 'Smoked Beer' SE MAPEARA CON UN ALGORITMO PERSONALIZADO PROPIO:

beer_reviews.csv: 'Smoked Beer'

BeerAdvocates 2019: 'Smoke Porter' o 'Smoke Beer'

3.1.- Si el 'beer_name' SI TIENE LA PALABRA 'Porter' se mapeara al estilo 'Smoke Porter'

3.2.- Si el 'beer_name' NO TIENE LA PALABRA 'Porter' se mapeara al estilo 'Smoke Beer'

##### 3.1.- Si el 'beer_name' SI TIENE LA PALABRA 'Porter' se mapeara al estilo 'Smoke Porter'

In [27]:
# HAY 2815 VALORACIONES DE CERVEZAS de 'beer_style' IGUAL A 'Smoked Beer'
#'Brettster,American Porter', 'American Porter',...
dfreviews[dfreviews['beer_style'] == 'Smoked Beer']['beer_style'].count()

2815

In [29]:
# HAY 204 VALORACIONES DE CERVEZAS de 'beer_style' IGUAL A 'Smoked Beer' Y 'beer_name' CON SUFIJO 'Porter'
#'Brettster,American Porter', 'American Porter',...
dfreviews[(dfreviews['beer_style'] == 'Smoked Beer') & \
          (dfreviews['beer_name'].str.contains('Porter'))]['beer_name'].count()

204

In [30]:
dfreviews.loc[(dfreviews['beer_style'] == 'Smoked Beer') & (dfreviews['beer_name'].str.contains('Porter')), \
              'beer_advocates_style'] = 'Smoke Porter'

In [32]:
dfreviews[(dfreviews['beer_style'] == 'Smoked Beer') & \
          (dfreviews['beer_advocates_style'] == 'Smoke Porter')] \
         [['beer_style', 'beer_name', 'beer_advocates_style']].head()

beer_style                   beer_name beer_advocates_style
7025   Smoked Beer   Simple Malt Double Porter         Smoke Porter
19033  Smoked Beer  Smokestack Heritage Porter         Smoke Porter
22455  Smoked Beer               Smoked Porter         Smoke Porter
23839  Smoked Beer  Smokestack Heritage Porter         Smoke Porter
24779  Smoked Beer  Smokestack Heritage Porter         Smoke Porter

In [33]:
# SE HAN MAPEADO 204 registros CORRECTAMENTE
dfreviews[(dfreviews['beer_style'] == 'Smoked Beer') & \
          (dfreviews['beer_advocates_style'] == 'Smoke Porter')]['beer_style'].count()

204

#### 3.2.- Si el 'beer_name' NO TIENE LA PALABRA 'Porter' se mapeara al estilo 'Smoke Beer' 

In [34]:
# HAY 2611 VALORACIONES DE CERVEZAS de 'beer_style' IGUAL A 'Smoked Beer' Y 'beer_name' SIN SUFIJO 'Porter'
#'Brettster,American Porter', 'American Porter,'
dfreviews[(dfreviews['beer_style'] == 'Smoked Beer') & \
          (dfreviews['beer_advocates_style'] == '') & \
          (not dfreviews['beer_name'].str.contains('Porter').all())]

brewery_id                          brewery_name  \
1344            590            New Glarus Brewing Company   
2165           1141                Brasserie Dieu Du Ciel   
3039            590            New Glarus Brewing Company   
4277          15711                       HaandBryggeriet   
5130           1471            Dark Horse Brewing Company   
5218            335           New Holland Brewing Company   
6257            132                            Rogue Ales   
6283           2958  BFM Brasserie Des Franches-Montagnes   
6355            345               Victory Brewing Company   
6455            898                       Heavy Seas Beer   
6527            335           New Holland Brewing Company   
7590            132                            Rogue Ales   
7906           1471            Dark Horse Brewing Company   
7928            335           New Holland Brewing Company   
8257          22150             Stillwater Artisanal Ales   
8578          22511                Hill Farmstead Brewery   
8599            898                       Heavy Seas Beer   
8684            392               Weyerbacher Brewing Co.   
8761           1471            Dark Horse Brewing Company   
9089            761                  CB's Brewing Company   
10636           132                            Rogue Ales   
10819           345               Victory Brewing Company   
11363           132                            Rogue Ales   
13153           392               Weyerbacher Brewing Co.   
13602           345               Victory Brewing Company   
13809            26    Three Floyds Brewing Co. & Brewpub   
13942           335           New Holland Brewing Company   
14327           132                            Rogue Ales   
16977          1471            Dark Horse Brewing Company   
17993           132                            Rogue Ales   
...             ...                                   ...   
1473708          26    Three Floyds Brewing Co. & Brewpub   
1476137         392               Weyerbacher Brewing Co.   
1477736       11031                    Brouwerij De Molen   
1478858        1471            Dark Horse Brewing Company   
1479346         392               Weyerbacher Brewing Co.   
1479553         590            New Glarus Brewing Company   
1479727       15711                       HaandBryggeriet   
1479927       16790  Brouwerij De Dochter van de Korenaar   
1481833       22618                   8 Wired Brewing Co.   
1481902        2845                  Invercargill Brewery   
1481903       11611                   Bridge Road Brewers   
1481905         132                            Rogue Ales   
1481962       15711                       HaandBryggeriet   
1481976       23246                          Yeastie Boys   
1482500        1141                Brasserie Dieu Du Ciel   
1482677        4778                      3 Ravens Brewing   
1482944         132                            Rogue Ales   
1485232         132                            Rogue Ales   
1485380         590            New Glarus Brewing Company   
1485812        2099            Church-Key Brewing Company   
1486176         590            New Glarus Brewing Company   
1486786       11207            Maui Brewing Co. (Brewpub)   
1487065        1141                Brasserie Dieu Du Ciel   
1487076        8950                 Brasserie Bièropholie   
1487660         132                            Rogue Ales   
1487823         392               Weyerbacher Brewing Co.   
1488901        1141                Brasserie Dieu Du Ciel   
1488998         132                            Rogue Ales   
1489234       20417                   White Birch Brewing   
1489297       22511                Hill Farmstead Brewery   

                 review_time  review_overall  review_aroma  review_appearance  \
1344     2011-02-21 23:35:31             3.5           4.0                3.5   
2165     2010-05-24 11:50:03             3.5           5.0      

In [35]:
dfreviews[(dfreviews['beer_style'] == 'Smoked Beer') & \
          (dfreviews['beer_advocates_style'] == '') & \
          (not dfreviews['beer_name'].str.contains('Porter').all())]['beer_style'].count()

2611

In [36]:
dfreviews.loc[((dfreviews['beer_style'] == 'Smoked Beer') & \
               (dfreviews['beer_advocates_style'] == '') & \
               (not dfreviews['beer_name'].str.contains('Porter').all())), \
              'beer_advocates_style'] = 'Smoke Beer'

In [38]:
dfreviews[(dfreviews['beer_style'] == 'Smoked Beer') & \
          (dfreviews['beer_advocates_style'] == 'Smoke Beer') & \
          (not dfreviews['beer_name'].str.contains('Porter').all())] \
         [['beer_style', 'beer_name', 'beer_advocates_style']].head()

beer_style                     beer_name beer_advocates_style
1344  Smoked Beer      Unplugged Smoked Rye Ale           Smoke Beer
2165  Smoked Beer                  Charbonniere           Smoke Beer
3039  Smoked Beer      Unplugged Smoked Rye Ale           Smoke Beer
4277  Smoked Beer                Norwegian Wood           Smoke Beer
5130  Smoked Beer  Dark Horse Fore Smoked Stout           Smoke Beer

In [39]:
# SE HAN MAPEADO 2611 registros CORRECTAMENTE
dfreviews[(dfreviews['beer_style'] == 'Smoked Beer') & \
          (dfreviews['beer_advocates_style'] == 'Smoke Beer') & \
          (not dfreviews['beer_name'].str.contains('Porter').all())] \
          ['beer_style'].count()

2611

#### 4.- 'Black & Tan' SE MAPEARA CON UN ALGORITMO PERSONALIZADO PROPIO

4.1.- Si el 'beer_name' SI TIENE LAS PALABRAS 'Black & Tan' O 'Black And Tan' se mapeara al estilo American Porter'

4.2.- Si el 'beer_name' NO TIENE LAS PALABRAS 'Black & Tan' NI 'Black And Tan' se mapeara a un estilo que dependera de su 'beer_name':

                                    Avg_ABV_RANGE ('abv_range' en ' beer_styles_v1.csv')
				'American Stout'	5.7-8.9%
				'American Porter'	4.0–7.0%
                
4.2.a.- Si 'beer_abv' >= 5.7 Y 'beer_abv' <= 8.9 se mapeara a 'American Stout'

4.2.b.- Si 'beer_abv' >= 4.0 Y 'beer_abv' <= 7.0 se mapeara a 'American Porter'

4.2.c.- Si 'beer_abv' < 4.0 se mapeara a 'American Porter'

4.2.d.- Si 'beer_abv' > 8.9 se mapeara a 'American Stout'

In [40]:
# 2183 VALORACIONES PENDIENTES DE MAPEAR PARA EL 'beer_style' DE VALOR 'Black & Tan'
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == '')]['beer_style'].count()

2183

In [41]:
# 42 'beer_name' DISTINTOS DEL 'beer_style' DE VALOR 'Black & Tan'
# grep 'Black & Tan' beer_reviews_v1.csv | cut -d',' -f11 | grep 'Tan' | sort
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == '')]['beer_name'] \
          .to_frame() \
          .sort_values(['beer_name'], ascending=False) \
          .drop_duplicates(subset= ['beer_name'], keep='first')['beer_name'].count()

42

#### 4.1.- Si el 'beer_name' SI TIENE LAS PALABRAS 'Black & Tan' O 'Black And Tan' se mapeara al estilo 'American Porter'

4.1.a.- HAY 1885 VALORACIONES DE CERVEZA DE 'beer_style' == 'Black & Tan' QUE PERTENECEN A 15 'beer_name' DISTINTAS QUE CONTIENEN 'Black & Tan' 

In [42]:
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == '') & \
          (dfreviews['beer_name'].str.contains('Black & Tan'))]['beer_name'].to_frame() \
         ['beer_name'].count()

1885

In [43]:
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == '') & \
          (dfreviews['beer_name'].str.contains('Black & Tan'))]['beer_name'].to_frame() \
          .sort_values(['beer_name'], ascending=False) \
          .drop_duplicates(subset= ['beer_name'], keep='first')['beer_name'].count()

15

In [44]:
dfreviews.loc[((dfreviews['beer_style'] == 'Black & Tan') & \
               (dfreviews['beer_advocates_style'] == '') & \
               (dfreviews['beer_name'].str.contains('Black & Tan'))), \
               'beer_advocates_style'] = 'American Porter'

In [45]:
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == 'American Porter')] \
         [['beer_style', 'beer_name', 'beer_advocates_style']].head()

beer_style                    beer_name beer_advocates_style
17    Black & Tan        Yuengling Black & Tan      American Porter
168   Black & Tan  Mississippi Mud Black & Tan      American Porter
3319  Black & Tan        Yuengling Black & Tan      American Porter
3905  Black & Tan          Saranac Black & Tan      American Porter
4406  Black & Tan         Michelob Black & Tan      American Porter

In [46]:
# SE HAN MAPEADO 1885 registros CORRECTAMENTE
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == 'American Porter')] \
         ['beer_style'].count()

1885

4.1.b.- HAY 16 VALORACIONES DE CERVEZA DE 'beer_style' == 'Black & Tan' QUE PERTENECEN A 3 'beer_name' DISTINTOS QUE CONTIENEN 'Black And Tan'

In [47]:
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == '') & \
          (dfreviews['beer_name'].str.contains('Black And Tan'))]['beer_name'].to_frame() \
         ['beer_name'].count()

16

In [48]:
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == '') & \
          (dfreviews['beer_name'].str.contains('Black And Tan'))]['beer_name'].to_frame() \
          .sort_values(['beer_name'], ascending=False) \
          .drop_duplicates(subset= ['beer_name'], keep='first')['beer_name'].count()

3

In [49]:
dfreviews.loc[((dfreviews['beer_style'] == 'Black & Tan') & \
               (dfreviews['beer_advocates_style'] == '') & \
               (dfreviews['beer_name'].str.contains('Black And Tan'))), \
               'beer_advocates_style'] = 'American Porter'

In [50]:
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == 'American Porter') & \
          (dfreviews['beer_name'].str.contains('Black And Tan'))] \
         [['beer_style', 'beer_name', 'beer_advocates_style']].head()

beer_style             beer_name beer_advocates_style
114806  Black & Tan  Gray's Black And Tan      American Porter
244365  Black & Tan         Black And Tan      American Porter
291069  Black & Tan  Gray's Black And Tan      American Porter
381432  Black & Tan  Gray's Black And Tan      American Porter
398657  Black & Tan         Black And Tan      American Porter

In [51]:
# SE HAN MAPEADO 16 registros CORRECTAMENTE
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == 'American Porter') & \
          (dfreviews['beer_name'].str.contains('Black And Tan'))] \
         [['beer_style', 'beer_name', 'beer_advocates_style']]['beer_style'].count()

16

#### 4.2.- Si el 'beer_name' NO TIENE LAS PALABRAS 'Black & Tan' NI 'Black And Tan' se mapeara a un estilo que dependera de su 'beer_name'.

HAY 282 VALORACIONES DE CERVEZA DE 'beer_style' == 'Black & Tan' QUE PERTENECEN A 24 'beer_name' DISTINTOS QUE NO CONTIENEN 'Black & Tan' NI CONTIENEN 'Black And Tan'

In [52]:
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == '') & \
          (not dfreviews['beer_name'].str.contains('Black & Tan').all()) & \
          (not dfreviews['beer_name'].str.contains('Black And Tan').all())] \
         ['beer_name'].to_frame()['beer_name'].count()

282

In [53]:
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == '') & \
          (not dfreviews['beer_name'].str.contains('Black & Tan').all()) & \
          (not dfreviews['beer_name'].str.contains('Black And Tan').all())] \
         ['beer_name'].to_frame() \
         .sort_values(['beer_name'], ascending=False) \
         .drop_duplicates(subset= ['beer_name'], keep='first') \
         ['beer_name'].count()

24

In [55]:
# 282 REGISTROS PARA MAPEAR A 'American Stout' o 'American Porter'
dfAux = dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
                  (dfreviews['beer_advocates_style'] == '') & \
                  (not dfreviews['beer_name'].str.contains('Black & Tan').all()) & \
                  (not dfreviews['beer_name'].str.contains('Black And Tan').all())]

In [56]:
print(dfAux['beer_style'] .count())
dfAux

282


brewery_id                           brewery_name  \
5365            200              Mendocino Brewing Company   
8126            200              Mendocino Brewing Company   
11268           200              Mendocino Brewing Company   
12576           200              Mendocino Brewing Company   
14808           200              Mendocino Brewing Company   
18183           200              Mendocino Brewing Company   
18603           200              Mendocino Brewing Company   
20961          2594                           Tree Brewing   
21268           823      The Olde Burnside Brewing Company   
32812           200              Mendocino Brewing Company   
34378           200              Mendocino Brewing Company   
45994           200              Mendocino Brewing Company   
57533           823      The Olde Burnside Brewing Company   
76511           200              Mendocino Brewing Company   
79456           200              Mendocino Brewing Company   
87307           200              Mendocino Brewing Company   
88303           200              Mendocino Brewing Company   
89468           823      The Olde Burnside Brewing Company   
97112           200              Mendocino Brewing Company   
118651          200              Mendocino Brewing Company   
126035          200              Mendocino Brewing Company   
130359          823      The Olde Burnside Brewing Company   
145266          823      The Olde Burnside Brewing Company   
145510          200              Mendocino Brewing Company   
153965          823      The Olde Burnside Brewing Company   
154110          823      The Olde Burnside Brewing Company   
163061          200              Mendocino Brewing Company   
170020          200              Mendocino Brewing Company   
172383          200              Mendocino Brewing Company   
178546          200              Mendocino Brewing Company   
...             ...                                    ...   
1326192         823      The Olde Burnside Brewing Company   
1327524         823      The Olde Burnside Brewing Company   
1345535         200              Mendocino Brewing Company   
1347225         823      The Olde Burnside Brewing Company   
1350547          33         Berkshire Brewing Company Inc.   
1360072         206  Hook & Ladder Brewing Company (No. 2)   
1375946         200              Mendocino Brewing Company   
1378673         200              Mendocino Brewing Company   
1384862         200              Mendocino Brewing Company   
1388771         200              Mendocino Brewing Company   
1391276         823      The Olde Burnside Brewing Company   
1391423         823      The Olde Burnside Brewing Company   
1397408         200              Mendocino Brewing Company   
1404505         823      The Olde Burnside Brewing Company   
1405105         200              Mendocino Brewing Company   
1411043        2594                           Tree Brewing   
1422250         200              Mendocino Brewing Company   
1432026         206  Hook & Ladder Brewing Company (No. 2)   
1448925         200              Mendocino Brewing Company   
1451077         206  Hook & Ladder Brewing Company (No. 2)   
1451846         200              Mendocino Brewing Company   
1452662        2594                           Tree Brewing   
1453280        2594                           Tree Brewing   
1461130         200              Mendocino Brewing Company   
1461626         206  Hook & Ladder Brewing Company (No. 2)   
1462023         823      The Olde Burnside Brewing Company   
1467106         823      The Olde Burnside Brewing Company   
1469944         200              Mendocino Brewing Company   
1483775         200              Mendocino Brewing Company   
1486734         200              Mendocino Brewing Company   

                 review_time  review_overall  review_aroma  review_appearance  \
5365     2007-03-24 23:58:25             4.0           3.0                3.5   
81

4.2.a.- Si 'beer_abv' >= 5.7 Y 'beer_abv' <= 8.9 se mapeara a 'American Stout'

In [63]:
# 161 REGISTROS DE MAPEO A 'American Stout'
dfAux[(dfAux['beer_abv'] >= 5.7) & (dfAux['beer_abv'] <= 8.9)] \
     [['beer_style', 'beer_name', 'beer_abv', 'beer_advocates_style']].head()

beer_style      beer_name  beer_abv beer_advocates_style
5365   Black & Tan  Black Eye Ale       6.5                     
8126   Black & Tan  Black Eye Ale       6.5                     
11268  Black & Tan  Black Eye Ale       6.5                     
12576  Black & Tan  Black Eye Ale       6.5                     
14808  Black & Tan  Black Eye Ale       6.5

In [58]:
dfAux[(dfAux['beer_abv'] >= 5.7) & (dfAux['beer_abv'] <= 8.9)]['beer_name'].count()

161

In [59]:
# 5 'beer_name' DE MAPEO A 'American Stout'
ldfAux = list(dfAux[(dfAux['beer_abv'] >= 5.7) & (dfAux['beer_abv'] <= 8.9)]['beer_name'].to_frame() \
                           .sort_values(['beer_name'], ascending=False) \
                           .drop_duplicates(subset= ['beer_name'], keep='first')['beer_name'])
ldfAux

['Flying Magic Zebras',
 'Dirty Harry',
 'Black Knight',
 'Black Eye Ale',
 'Berkshire Bourbon Shabadoo']

In [60]:
dfreviews.loc[dfreviews.apply((lambda row: row['beer_name'] in ldfAux), axis =1), \
              'beer_advocates_style'] = 'American Stout'

In [65]:
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == 'American Stout') & \
          (not dfreviews['beer_name'].str.contains('Black & Tan').all()) & \
          (not dfreviews['beer_name'].str.contains('Black And Tan').all()) & \
          (dfreviews.apply((lambda row: row['beer_name'] in ldfAux), axis =1))] \
         [['beer_style', 'beer_name', 'beer_abv', 'beer_advocates_style']].head()

beer_style      beer_name  beer_abv beer_advocates_style
5365   Black & Tan  Black Eye Ale       6.5       American Stout
8126   Black & Tan  Black Eye Ale       6.5       American Stout
11268  Black & Tan  Black Eye Ale       6.5       American Stout
12576  Black & Tan  Black Eye Ale       6.5       American Stout
14808  Black & Tan  Black Eye Ale       6.5       American Stout

In [66]:
# SE HAN MAPEADO 161 registros CORRECTAMENTE
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == 'American Stout') & \
          (not dfreviews['beer_name'].str.contains('Black & Tan').all()) & \
          (not dfreviews['beer_name'].str.contains('Black And Tan').all()) & \
          (dfreviews.apply((lambda row: row['beer_name'] in ldfAux), axis =1))] \
         ['beer_name'].count()

161

4.2.b.- Si 'beer_abv' >= 4.0 Y 'beer_abv' <= 7.0 se mapeara a 'American Porter'

In [70]:
# TODOS LOS REGISTROS QUE QUEDAN POR MAPEAR
dfAux = dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
                  (dfreviews['beer_advocates_style'] == '') & \
                  (not dfreviews['beer_name'].str.contains('Black & Tan').all()) & \
                  (not dfreviews['beer_name'].str.contains('Black And Tan').all())]

In [71]:
# 114 REGISTROS DE MAPEO A 'American Porter'
dfAux[(dfAux['beer_abv'] >= 4.0) & (dfAux['beer_abv'] <= 7.0)] \
     [['beer_style', 'beer_name', 'beer_abv', 'beer_advocates_style']].head()

beer_style            beer_name  beer_abv beer_advocates_style
20961   Black & Tan  Black Tree Dark Ale       5.0                     
57533   Black & Tan      Dirty Penny Ale       5.0                     
89468   Black & Tan      Dirty Penny Ale       5.0                     
130359  Black & Tan      Dirty Penny Ale       5.0                     
145266  Black & Tan      Dirty Penny Ale       5.0

In [72]:
dfAux[(dfAux['beer_abv'] >= 4.0) & (dfAux['beer_abv'] <= 7.0)]['beer_name'].count()

114

In [73]:
# 15 'beer_name' DE MAPEO A 'American Porter'
ldfAux = list(dfAux[(dfAux['beer_abv'] >= 4.0) & (dfAux['beer_abv'] <= 7.0)]['beer_name'].to_frame() \
                           .sort_values(['beer_name'], ascending=False) \
                           .drop_duplicates(subset= ['beer_name'], keep='first')['beer_name'])

print(len(ldfAux))
ldfAux

15


['Åbro Bryggmästarens Mixed Ale 5,0%',
 "Zlatopramen Half N' Half",
 'Whytemud Whysky Lager',
 'Table Beer (Half And Half)',
 'Spendrups Fifty Fifty',
 'Serendipity No. 1',
 "Salm Bräu G'Mischtes",
 'Hook & Ladder Ember Amber',
 'Highland Piper Tartan Ale',
 'Half & Half',
 'French Quarter Smoked Chipotle',
 'Dirty Penny Ale',
 'Blackout Ale',
 'BlackFish',
 'Black Tree Dark Ale']

In [74]:
dfreviews.loc[dfreviews.apply((lambda row: row['beer_name'] in ldfAux), axis =1), \
              'beer_advocates_style'] = 'American Porter'

In [75]:
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == 'American Porter') & \
          (not dfreviews['beer_name'].str.contains('Black & Tan').all()) & \
          (not dfreviews['beer_name'].str.contains('Black And Tan').all()) & \
          (dfreviews.apply((lambda row: row['beer_name'] in ldfAux), axis =1))] \
         [['beer_style', 'beer_name', 'beer_abv', 'beer_advocates_style']].head()

beer_style            beer_name  beer_abv beer_advocates_style
20961   Black & Tan  Black Tree Dark Ale       5.0      American Porter
57533   Black & Tan      Dirty Penny Ale       5.0      American Porter
89468   Black & Tan      Dirty Penny Ale       5.0      American Porter
130359  Black & Tan      Dirty Penny Ale       5.0      American Porter
145266  Black & Tan      Dirty Penny Ale       5.0      American Porter

In [76]:
# SE HAN MAPEADO 114 registros CORRECTAMENTE
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == 'American Porter') & \
          (not dfreviews['beer_name'].str.contains('Black & Tan').all()) & \
          (not dfreviews['beer_name'].str.contains('Black And Tan').all()) & \
          (dfreviews.apply((lambda row: row['beer_name'] in ldfAux), axis =1))] \
         ['beer_name'].count()

114

4.2.c.- Si 'beer_abv' < 4.0 se mapeara a 'American Porter'

In [77]:
# TODOS LOS REGISTROS QUE QUEDAN POR MAPEAR
dfAux = dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
                  (dfreviews['beer_advocates_style'] == '') & \
                  (not dfreviews['beer_name'].str.contains('Black & Tan').all()) & \
                  (not dfreviews['beer_name'].str.contains('Black And Tan').all())]

In [78]:
# 01 REGISTRO DE MAPEO A 'American Porter'
dfAux[dfAux['beer_abv'] < 4.0] \
     [['beer_style', 'beer_name', 'beer_abv', 'beer_advocates_style']].head()

beer_style                           beer_name  beer_abv  \
785822  Black & Tan  Åbro Bryggmästarens Mixed Ale 3,5%       3.5   

       beer_advocates_style  
785822

In [79]:
# 01 'beer_name' DE MAPEO A 'American Porter'
ldfAux = list(dfAux[dfAux['beer_abv'] < 4.0]['beer_name'])
ldfAux

['Åbro Bryggmästarens Mixed Ale 3,5%']

In [80]:
dfreviews.loc[dfreviews.apply((lambda row: row['beer_name'] in ldfAux), axis =1), \
              'beer_advocates_style'] = 'American Porter'

In [81]:
# SE HAN MAPEADO 01 registros CORRECTAMENTE
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == 'American Porter') & \
          (not dfreviews['beer_name'].str.contains('Black & Tan').all()) & \
          (not dfreviews['beer_name'].str.contains('Black And Tan').all()) & \
          (dfreviews.apply((lambda row: row['beer_name'] in ldfAux), axis =1))]

brewery_id   brewery_name          review_time  review_overall  \
785822        3825  Åbro Bryggeri  2005-07-12 18:21:48             3.0   

        review_aroma  review_appearance review_profilename   beer_style  \
785822           3.0                3.5               bark  Black & Tan   

        review_palate  review_taste                           beer_name  \
785822            3.0           3.0  Åbro Bryggmästarens Mixed Ale 3,5%   

        beer_abv  beer_beerid  review_average  abv_strength  \
785822       3.5        23685             3.1             1   

       beer_advocates_style  
785822      American Porter

4.2.d.- Si 'beer_abv' > 8.9 se mapeara a 'American Stout

In [82]:
# TODOS LOS REGISTROS QUE QUEDAN POR MAPEAR
dfAux = dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
                  (dfreviews['beer_advocates_style'] == '') & \
                  (not dfreviews['beer_name'].str.contains('Black & Tan').all()) & \
                  (not dfreviews['beer_name'].str.contains('Black And Tan').all())]

In [84]:
# 06 REGISTROS DE MAPEO A 'American Stout'
dfAux[dfAux['beer_abv'] > 8.9] \
     [['beer_style', 'beer_name', 'beer_abv', 'beer_advocates_style']].head()

beer_style                            beer_name  beer_abv  \
21268    Black & Tan                     Stone Of Destiny      12.0   
437133   Black & Tan  Imperial Black Widow Hop Head Stout      10.0   
903539   Black & Tan               Birkenhead Black Snake      10.0   
991544   Black & Tan                     Stone Of Destiny      12.0   
1022752  Black & Tan                     Stone Of Destiny      12.0   

        beer_advocates_style  
21268                         
437133                        
903539                        
991544                        
1022752

In [85]:
dfAux[dfAux['beer_abv'] > 8.9]['beer_name'].count()

6

In [86]:
# 3 'beer_name' DE MAPEO A 'American Stout'
ldfAux = list(dfAux[dfAux['beer_abv'] > 8.9]['beer_name'].to_frame() \
              .sort_values(['beer_name'], ascending=False) \
              .drop_duplicates(subset= ['beer_name'], keep='first')['beer_name'])

print(len(ldfAux))
ldfAux

3


['Stone Of Destiny',
 'Imperial Black Widow Hop Head Stout',
 'Birkenhead Black Snake']

In [89]:
dfreviews[dfreviews.apply((lambda row: row['beer_name'] in ldfAux), axis =1)] \
         [['beer_style', 'beer_name', 'beer_abv', 'beer_advocates_style']]

beer_style                            beer_name  beer_abv  \
21268    Black & Tan                     Stone Of Destiny      12.0   
437133   Black & Tan  Imperial Black Widow Hop Head Stout      10.0   
903539   Black & Tan               Birkenhead Black Snake      10.0   
991544   Black & Tan                     Stone Of Destiny      12.0   
1022752  Black & Tan                     Stone Of Destiny      12.0   
1038862  Black & Tan                     Stone Of Destiny      12.0   

        beer_advocates_style  
21268                         
437133                        
903539                        
991544                        
1022752                       
1038862

In [90]:
dfreviews.loc[dfreviews.apply((lambda row: row['beer_name'] in ldfAux), axis =1), \
              'beer_advocates_style'] = 'American Stout'

In [91]:
# SE HAN MAPEADO 06 registros CORRECTAMENTE
dfreviews[(dfreviews['beer_style'] == 'Black & Tan') & \
          (dfreviews['beer_advocates_style'] == 'American Stout') & \
          (not dfreviews['beer_name'].str.contains('Black & Tan').all()) & \
          (not dfreviews['beer_name'].str.contains('Black And Tan').all()) & \
          (dfreviews.apply((lambda row: row['beer_name'] in ldfAux), axis =1))] \
         [['beer_style', 'beer_name', 'beer_abv', 'beer_advocates_style']]

beer_style                            beer_name  beer_abv  \
21268    Black & Tan                     Stone Of Destiny      12.0   
437133   Black & Tan  Imperial Black Widow Hop Head Stout      10.0   
903539   Black & Tan               Birkenhead Black Snake      10.0   
991544   Black & Tan                     Stone Of Destiny      12.0   
1022752  Black & Tan                     Stone Of Destiny      12.0   
1038862  Black & Tan                     Stone Of Destiny      12.0   

        beer_advocates_style  
21268         American Stout  
437133        American Stout  
903539        American Stout  
991544        American Stout  
1022752       American Stout  
1038862       American Stout

#### NOS QUEDAN 1489510 - 962747 = 526763 VALORACIONES CUYO 'beer_style' NO ESTA MAPEADO TODAVIA PARA 41 'beer_style' DISTINTOS

In [92]:
dfreviews[dfreviews['beer_advocates_style'] != '']['beer_advocates_style'].count()

962747

In [94]:
print((dfreviews['beer_style'].count() - \
       dfreviews[dfreviews['beer_advocates_style'] != '']['beer_advocates_style'].count()))

526763


In [95]:
dfreviews[dfreviews['beer_advocates_style'] == '']['beer_style'].to_frame() \
          .sort_values(['beer_style'], ascending=False) \
          .drop_duplicates(subset= ['beer_style'], keep='first')['beer_style'].count()

41

In [97]:
dfreviews[dfreviews['beer_advocates_style'] == '']['beer_style'].to_frame() \
          .sort_values(['beer_style'], ascending=False) \
          .drop_duplicates(subset= ['beer_style'], keep='first')

beer_style
758201                             Witbier
1472884                          Wheatwine
576551                          Weizenbock
639893                              Tripel
1076742                        Schwarzbier
1340822                              Sahti
389028                          Roggenbier
10742                            Rauchbier
576666                    Quadrupel (Quad)
1141223                            Old Ale
710654                       Oatmeal Stout
838267                  Märzen Oktoberfest
923099                    Milk Sweet Stout
1388706                        Light Lager
120979                        Lambic Fruit
623895                              Kölsch
929403                               Kvass
139052                       Kristalweizen
1025927                         Hefeweizen
86690                             Happoshu
1317582                             Gueuze
656457                                Gose
358509                Fruit Vegetable Beer
119437                                Faro
1155933  Extra Special Strong Bitter (ESB)
714178                   Euro Strong Lager
1195562                    Euro Pale Lager
80709                      Euro Dark Lager
477531                             Eisbock
646725                        Dunkelweizen
1039348                             Dubbel
698511             Dortmunder Export Lager
525048                          Doppelbock
352269                           Cream Ale
204064                                Bock
37892                       Bière de Garde
1143268                 Berliner Weissbier
191631      American Double Imperial Stout
1107460   American Double Imperial Pilsner
1437997       American Double Imperial IPA
1204318                            Altbier

#### 5.- MAPEO CON UN ALGORITMO COMUN PARA EL RESTO DE LOS 'beer_style'

In [99]:
def words_number_found(input_word, words_to_search_for_list, separator_char):
    words_of_input_word = input_word.split(separator_char)
    #print('words_of_input_word: "{0}"'.format(words_of_input_word))
    counters_dict = {}
    for word_to_search_for in words_to_search_for_list:
        #print('word_to_search_for: "{0}"'.format(word_to_search_for))
        words_of_word_to_search_for = word_to_search_for.split(separator_char)
        found_words_counter = 0
        #print('words_of_word_to_search_for: "{0}"'.format(words_of_word_to_search_for))
        for word in words_of_input_word:
            if word in words_of_word_to_search_for:
                found_words_counter += 1
        #print('found_words_counter: "{0}"'.format(found_words_counter))
        counters_dict[word_to_search_for] = found_words_counter

    return counters_dict

In [100]:
def mapping_beer_style_not_found(bs, bslist):
    bsd = words_number_found(bs, bslist, ' ')
    found_bs = ''
    max_bs_counter = 0
    for bs in bsd:
        if bsd[bs] > max_bs_counter:
            max_bs_counter = bsd[bs]
            found_bs = bs
    return found_bs

In [101]:
bs_mapped_df = dfreviews[dfreviews['beer_advocates_style'] == '']['beer_style'].to_frame() \
               .sort_values(['beer_style'], ascending=False) \
               .drop_duplicates(subset= ['beer_style'], keep='first')['beer_style'].to_frame()

In [102]:
print(bs_mapped_df.count())
print(type(bs_mapped_df))
bs_mapped_df

beer_style    41
dtype: int64
<class 'pandas.core.frame.DataFrame'>


beer_style
758201                             Witbier
1472884                          Wheatwine
576551                          Weizenbock
639893                              Tripel
1076742                        Schwarzbier
1340822                              Sahti
389028                          Roggenbier
10742                            Rauchbier
576666                    Quadrupel (Quad)
1141223                            Old Ale
710654                       Oatmeal Stout
838267                  Märzen Oktoberfest
923099                    Milk Sweet Stout
1388706                        Light Lager
120979                        Lambic Fruit
623895                              Kölsch
929403                               Kvass
139052                       Kristalweizen
1025927                         Hefeweizen
86690                             Happoshu
1317582                             Gueuze
656457                                Gose
358509                Fruit Vegetable Beer
119437                                Faro
1155933  Extra Special Strong Bitter (ESB)
714178                   Euro Strong Lager
1195562                    Euro Pale Lager
80709                      Euro Dark Lager
477531                             Eisbock
646725                        Dunkelweizen
1039348                             Dubbel
698511             Dortmunder Export Lager
525048                          Doppelbock
352269                           Cream Ale
204064                                Bock
37892                       Bière de Garde
1143268                 Berliner Weissbier
191631      American Double Imperial Stout
1107460   American Double Imperial Pilsner
1437997       American Double Imperial IPA
1204318                            Altbier

Recordemos nuestra lista de 'style_name' de ESTILOS DE CERVEZAS DE 2019 DE BeerAdvocates

In [103]:
bslist

['German Bock',
 'German Doppelbock',
 'German Eisbock',
 'German Maibock',
 'German Weizenbock',
 'American Brown Ale',
 'English Brown Ale',
 'English Dark Mild Ale',
 'German Altbier',
 'American Black Ale',
 'Belgian Dark Ale',
 'Belgian Dubbel',
 'German Roggenbier',
 'Scottish Ale',
 'Winter Warmer',
 'American Amber Red Lager',
 'European Dark Lager',
 'German Märzen Oktoberfest',
 'German Rauchbier',
 'German Schwarzbier',
 'Munich Dunkel Lager',
 'Vienna Lager',
 'American Cream Ale',
 'Bière de Champagne Bière Brut',
 'Braggot',
 'California Common Steam Beer',
 'American Brut IPA',
 'American Imperial IPA',
 'American IPA',
 'Belgian IPA',
 'English India Pale Ale (IPA)',
 'New England IPA',
 'American Amber Red Ale',
 'American Blonde Ale',
 'American Pale Ale (APA)',
 'Belgian Blonde Ale',
 'Belgian Pale Ale',
 'Belgian Saison',
 'English Bitter',
 'English Extra Special Strong Bitter (ESB)',
 'English Pale Ale',
 'English Pale Mild Ale',
 'French Bière de Garde',
 'German

COMPROBACION PREVIA ANTES DE LA SEGUNDA EJECUCION DE mapping_beer_style_not_found()

In [104]:
bs_mapped_df['beer_advocates_style'] = bs_mapped_df['beer_style'] \
        .apply(lambda beer_style: mapping_beer_style_not_found(beer_style, bslist))

bs_mapped_df

beer_style  \
758201                             Witbier   
1472884                          Wheatwine   
576551                          Weizenbock   
639893                              Tripel   
1076742                        Schwarzbier   
1340822                              Sahti   
389028                          Roggenbier   
10742                            Rauchbier   
576666                    Quadrupel (Quad)   
1141223                            Old Ale   
710654                       Oatmeal Stout   
838267                  Märzen Oktoberfest   
923099                    Milk Sweet Stout   
1388706                        Light Lager   
120979                        Lambic Fruit   
623895                              Kölsch   
929403                               Kvass   
139052                       Kristalweizen   
1025927                         Hefeweizen   
86690                             Happoshu   
1317582                             Gueuze   
656457                                Gose   
358509                Fruit Vegetable Beer   
119437                                Faro   
1155933  Extra Special Strong Bitter (ESB)   
714178                   Euro Strong Lager   
1195562                    Euro Pale Lager   
80709                      Euro Dark Lager   
477531                             Eisbock   
646725                        Dunkelweizen   
1039348                             Dubbel   
698511             Dortmunder Export Lager   
525048                          Doppelbock   
352269                           Cream Ale   
204064                                Bock   
37892                       Bière de Garde   
1143268                 Berliner Weissbier   
191631      American Double Imperial Stout   
1107460   American Double Imperial Pilsner   
1437997       American Double Imperial IPA   
1204318                            Altbier   

                              beer_advocates_style  
758201                             Belgian Witbier  
1472884                     American Wheatwine Ale  
576551                           German Weizenbock  
639893                              Belgian Tripel  
1076742                         German Schwarzbier  
1340822                              Finnish Sahti  
389028                           German Roggenbier  
10742                             German Rauchbier  
576666                    Belgian Quadrupel (Quad)  
1141223                            English Old Ale  
710654                       English Oatmeal Stout  
838267                   German Märzen Oktoberfest  
923099                    English Sweet Milk Stout  
1388706                       American Light Lager  
120979                        Belgian Fruit Lambic  
623895                               German Kölsch  
929403                               Russian Kvass  
139052                        German Kristalweizen  
1025927                          German Hefeweizen  
86690                            Japanese Happoshu  
1317582                             Belgian Gueuze  
656457                                Leipzig Gose  
358509                        Fruit and Field Beer  
119437                                Belgian Faro  
1155933  English Extra Special Strong Bitter (ESB)  
714178                       European Strong Lager  
1195562                        European Pale Lager  
80709                          European Dark Lager  
477531                              German Eisbock  
646725                         German Dunkelweizen  
1039348                             Belgian Dubbel  
698511                  European Export Dortmunder  
525048                           German Doppelbock  
352269                          American Cream Ale  
204064                                 German Bock  
37892                        French Bière de Garde  
1143268                            Berliner Weisse  
191631                     American Imperial Stout  
1107460                  America

EJECUCION DEL ALGORITMO DE MAPEO COMUN mapping_beer_style_not_found() PAR MAPEAR LAS 526763 VALORACIONES RESTANTES

In [105]:
bs_to_automatically_map_list = list(bs_mapped_df['beer_style'])

print(len(bs_to_automatically_map_list))
bs_to_automatically_map_list

41


['Witbier',
 'Wheatwine',
 'Weizenbock',
 'Tripel',
 'Schwarzbier',
 'Sahti',
 'Roggenbier',
 'Rauchbier',
 'Quadrupel (Quad)',
 'Old Ale',
 'Oatmeal Stout',
 'Märzen Oktoberfest',
 'Milk Sweet Stout',
 'Light Lager',
 'Lambic Fruit',
 'Kölsch',
 'Kvass',
 'Kristalweizen',
 'Hefeweizen',
 'Happoshu',
 'Gueuze',
 'Gose',
 'Fruit Vegetable Beer',
 'Faro',
 'Extra Special Strong Bitter (ESB)',
 'Euro Strong Lager',
 'Euro Pale Lager',
 'Euro Dark Lager',
 'Eisbock',
 'Dunkelweizen',
 'Dubbel',
 'Dortmunder Export Lager',
 'Doppelbock',
 'Cream Ale',
 'Bock',
 'Bière de Garde',
 'Berliner Weissbier',
 'American Double Imperial Stout',
 'American Double Imperial Pilsner',
 'American Double Imperial IPA',
 'Altbier']

In [106]:
dfreviews[dfreviews['beer_style'] \
          .apply(lambda beer_style: beer_style in bs_to_automatically_map_list)]['beer_style'].count()

526763

In [107]:
for bs in bs_to_automatically_map_list:
    dfreviews.loc[dfreviews['beer_style'] == bs, 'beer_advocates_style'] = \
        mapping_beer_style_not_found(bs, bslist)

In [108]:
dfreviews[dfreviews['beer_style'] \
          .apply(lambda beer_style: beer_style in bs_to_automatically_map_list)] \
         [['beer_style', 'beer_name', 'beer_advocates_style']].head(10)

beer_style  \
3                             Altbier   
5                Fruit Vegetable Beer   
6                                Bock   
7   Extra Special Strong Bitter (ESB)   
11                             Dubbel   
16                        Light Lager   
18                        Schwarzbier   
20                          Cream Ale   
21                       Dunkelweizen   
23     American Double Imperial Stout   

                                   beer_name  \
3                                 Double Bag   
5                           Éphémère (Apple)   
6                             Yuengling Bock   
7      Sierra Nevada ESB (Early Spring Beer)   
11  La Trappe Dubbel (Koningshoeven Dominus)   
16                     Yuengling Light Lager   
18                          Xingu Black Beer   
20                   Wexford Irish Cream Ale   
21                Tucher Dunkles Hefe Weizen   
23                                 The Abyss   

                         beer_advocates_style  
3                              German Altbier  
5                        Fruit and Field Beer  
6                                 German Bock  
7   English Extra Special Strong Bitter (ESB)  
11                             Belgian Dubbel  
16                       American Light Lager  
18                         German Schwarzbier  
20                         American Cream Ale  
21                        German Dunkelweizen  
23                    American Imperial Stout

In [109]:
# SE HAN MAPEADO 526763 registros CORRECTAMENTE
dfreviews[dfreviews['beer_style'] \
          .apply(lambda beer_style: beer_style in bs_to_automatically_map_list)] \
         ['beer_style'].count()

526763

#### SO, THAT'S ALL FOLKS

In [110]:
dfreviews[dfreviews['beer_advocates_style'] == '']['beer_style'].count()

0

#### Almacenaremos el DATASET CON LOS 'beer_style' DE 2015 DE BeerAdvocates MAPEADOS A LOS 'style_name' DE 2019

In [111]:
dfreviews.to_csv('./Data/beer_reviews_v1_1.csv', sep=',', index=False)